src/transforms/projectionpursuit.jl

```diff
@@ -0,0 +1,196 @@
+
+The non-singularity of Q is controlled by assuring that norm(det(Q)) ≥ `tol`. The iterative 
+process terminates whenever the transformed samples are "more Gaussian" than `perc`% of `n`
+randomly generated samples from the standard multivariate Gaussian distribution, or when the 
+number of iterations reaches a maximum `maxiter`.
+
+# Examples
+
+```julia
+ProjectionPursuit()
+ProjectionPursuit(deg=10)
+ProjectionPursuit(perc=.85, n=50)
+ProjectionPursuit(tol=1e-4, maxiter=250, deg=5, perc=.95, n=100)
+```
+
+"""
+
+struct ProjectionPursuit{T} <: Stateless
+  tol::T
+  maxiter::Int
+  deg::Int
+  perc::T
+  n::Int
+end
+
+ProjectionPursuit(;tol=1e-6, maxiter=100, deg=5, perc=.95, n=100) =
+  ProjectionPursuit{typeof(tol)}(tol, maxiter, deg, perc, n)
+
+isrevertible(::Type{<:ProjectionPursuit}) = true
+
+# transforms a row of random variables into a convex combination 
+# of random variables with values in [-1,1] and standard normal distribution
+function rscore(Z, α)
+  ᾱ = (1/norm(α)) .* α
+  X = Z * ᾱ
+  2 .* cdf.(Normal(), X) .- 1
+end
+
+# projection index of sample along a given direction
+function pindex(transform, Z, α)
+  X = rscore(Z, α)
+  I = (3/2) * mean(X)^2
+  if transform.deg > 1
+    Pⱼ₋₂, Pⱼ₋₁ = ones(size(X)), X
+    for j = 2:transform.deg
+      Pⱼ₋₂, Pⱼ₋₁ = 
+        Pⱼ₋₁, (1/j) * ((2j-1) * X .* Pⱼ₋₁ - (j-1) * Pⱼ₋₂)
+      I += ((2j+1)/2) * (mean(Pⱼ₋₁))^2
+    end
+  end
+  I
+end
+
+function optimaxis(transform, Z)
+  _, q = size(Z)
+  E = 1.0 * I(q)
+  f(i) = pindex(transform, Z, E[:,i])
+  Iₘₐₓ, iₘₐₓ = findmax(f, 1:q)
+  E[:,iₘₐₓ]
+end
+
+function optimdiag(transform, Z, α₀)
+  q  = length(α₀)
+  E  = 1.0 * I(q)
+  Iₐ = pindex(transform, Z, α₀)
+  α  = α₀
+  for i in 1:q
+    eᵢ = E[:,i]
+    f₊ = pindex(transform, Z, (1/√(2+2α[i])) * (α + eᵢ))
+    f₋ = pindex(transform, Z, (1/√(2-2α[i])) * (α - eᵢ)) * (α[i] != 1)
+    f = (f₊ > f₋) * f₊ + (f₊ ≤ f₋) * f₋ 
+    s = (f₊ > f₋) * 1 + (f₊ ≤ f₋) * (-1)
+    if f > Iₐ
+      Iₐ = f
+      α  = (1/√(2+2s*α[i])) * (α + s*eᵢ)
+    end
+  end
+  α
+end
+
+function optimall(transform, Z, α₀)
+  f(α) = -pindex(transform, Z, α)
+  op = optimize(f, α₀)
+  minimizer(op)
+end
+
+function optim(transform, Z)
+  α₁ = optimaxis(transform, Z)
+  α₂ = optimdiag(transform, Z, α₁)
+  optimall(transform, Z, α₂)  
+end
+
+function rmstructure(transform, Z, α)
+  q = length(α)
+  Q, R = qr(hcat(α, rand(q, q-1)))
+  while norm(diag(R)) < transform.tol
+    Q, R = qr(hcat(α, rand(q, q-1)))
+  end
+  T = Z * Q
+  table = Tables.table(T)
+  newtable, colcache = apply(Quantile(), table)
+  T[:,1] = Tables.getcolumn(newtable, 1)
+  T * Q', Q, colcache
+end
+
+# projection index of the standard multivariate Gaussian
+function gausspmatrix(transform, N, q)
+  n  = transform.n
+  E  = 1.0 * I(q)
+  Iₙ = Matrix{Float64}(undef, n, q)
+  for i = 1:n-1
+    S = randn(N, q)
+    for j = 1:q
+      Iₙ[i,j] = pindex(transform, S, E[:,j])
+    end
+  end
+  Iₙ
+end
```

Comment:
Let's rename this function to gindex and move it to after the function pindex so that readers understand the dimensions of the arrays early.

Reply:
Okay. I was even thinking about doing the quantile process inside this function and not in apply().

Reply:
The idea is interesting, we should look at it again after we address the other issues.